File: backend/app/schemas/analytics.py
"""Pydantic schemas for repository analytics API."""

from __future__ import annotations

from pydantic import BaseModel, Field


class SymbolRefResponse(BaseModel):
    name: str
    file_path: str
    start_line: int
    end_line: int
    kind: str
    lines: int = 0
    complexity_score: int = 0


class DuplicateClusterResponse(BaseModel):
    fingerprint: str
    count: int
    similarity: float
    symbols: list[SymbolRefResponse]


class LanguageSliceResponse(BaseModel):
    language: str
    files: int
    lines: int
    percentage: float


class ComplexitySliceResponse(BaseModel):
    level: str
    count: int


class HeatmapCellResponse(BaseModel):
    path: str
    directory: str
    language: str
    lines: int
    complexity_score: int
    symbol_count: int
    complexity_label: str


class DependencyNodeResponse(BaseModel):
    id: str
    label: str
    path: str
    group: str


class DependencyEdgeResponse(BaseModel):
    source: str
    target: str
    label: str


class DependencyGraphResponse(BaseModel):
    nodes: list[DependencyNodeResponse]
    edges: list[DependencyEdgeResponse]
    max_depth: int


class RepositoryAnalyticsResponse(BaseModel):
    repo_id: str
    job_id: str
    file_count: int
    function_count: int
    class_count: int
    dependency_depth: int
    most_complex_file: SymbolRefResponse | None = None
    largest_function: SymbolRefResponse | None = None
    duplicate_clusters: list[DuplicateClusterResponse] = Field(default_factory=list)
    duplicate_estimate: int = 0
    dead_code_symbols: list[SymbolRefResponse] = Field(default_factory=list)
    dead_code_estimate: int = 0
    language_distribution: list[LanguageSliceResponse] = Field(default_factory=list)
    complexity_distribution: list[ComplexitySliceResponse] = Field(default_factory=list)
    heatmap: list[HeatmapCellResponse] = Field(default_factory=list)
    dependency_graph: DependencyGraphResponse
    summary: dict[str, int | float | str | None] = Field(default_factory=dict)

File: frontend/src/types/analytics.ts
export interface SymbolRef {
  name: string;
  file_path: string;
  start_line: number;
  end_line: number;
  kind: string;
  lines: number;
  complexity_score: number;
}

export interface DuplicateCluster {
  fingerprint: string;
  count: number;
  similarity: number;
  symbols: SymbolRef[];
}

export interface LanguageSlice {
  language: string;
  files: number;
  lines: number;
  percentage: number;
}

export interface ComplexitySlice {
  level: string;
  count: number;
}

export interface HeatmapCell {
  path: string;
  directory: string;
  language: string;
  lines: number;
  complexity_score: number;
  symbol_count: number;
  complexity_label: string;
}

export interface DependencyNode {
  id: string;
  label: string;
  path: string;
  group: string;
}

export interface DependencyEdge {
  source: string;
  target: string;
  label: string;
}

export interface DependencyGraph {
  nodes: DependencyNode[];
  edges: DependencyEdge[];
  max_depth: number;
}

export interface RepositoryAnalytics {
  repo_id: string;
  job_id: string;
  file_count: number;
  function_count: number;
  class_count: number;
  dependency_depth: number;
  most_complex_file: SymbolRef | null;
  largest_function: SymbolRef | null;
  duplicate_clusters: DuplicateCluster[];
  duplicate_estimate: number;
  dead_code_symbols: SymbolRef[];
  dead_code_estimate: number;
  language_distribution: LanguageSlice[];
  complexity_distribution: ComplexitySlice[];
  heatmap: HeatmapCell[];
  dependency_graph: DependencyGraph;
  summary: Record<string, number | string | null>;
}
